waku/v2/peerstore/waku_peer_store.go

```diff
@@ -207,7 +208,38 @@ func (ps *WakuPeerstoreImpl) PubSubTopics(p peer.ID) ([]string, error) {
 	return result.([]string), nil
 }
 
-// PeersByPubSubTopic Returns list of peers by pubSubTopic
+// PeersByPubSubTopic Returns list of peers that support list of pubSubTopics
+// If specifiPeers are listed, filtering is done from them otherwise from all peers in peerstore
+func (ps *WakuPeerstoreImpl) PeersByPubSubTopics(pubSubTopics []string, specificPeers ...peer.ID) peer.IDSlice {
+	if specificPeers == nil {
+		specificPeers = ps.Peers()
+	}
+	var result peer.IDSlice
+	for _, p := range specificPeers {
+		topics, err := ps.PubSubTopics(p)
+		if err == nil {
+			//Convoluted and crazy logic to find subset of topics
+			// Could not find a better way to do it?
+			peerTopicMap := make(map[string]struct{})
+			match := true
+			for _, topic := range topics {
+				peerTopicMap[topic] = struct{}{}
+			}
+			for _, topic := range pubSubTopics {
+				if _, ok := peerTopicMap[topic]; !ok {
+					match = false
+					break
+				}
+			}
+			if match {
+				result = append(result, p)
```

Comment:
Could you not store a shard set for each peers then use set operations?

There are probably fancier way to store this data for easy access but would be good enough IMO.